storage/badger/dkg_keys.go

```diff
@@ -66,12 +66,30 @@ func (k *DKGKeys) retrieveTx(epochCounter uint64) func(tx *badger.Txn) (*dkg.DKG
 	}
 }
 
+// InsertMyDKGPrivateInfo insert the DKG private key to database when DKG was completed, and a DKG private key was successfully generated.
 func (k *DKGKeys) InsertMyDKGPrivateInfo(epochCounter uint64, info *dkg.DKGParticipantPriv) error {
 	return operation.RetryOnConflictTx(k.db, transaction.Update, k.storeTx(epochCounter, info))
 }
 
-func (k *DKGKeys) RetrieveMyDKGPrivateInfo(epochCounter uint64) (*dkg.DKGParticipantPriv, error) {
+// InsertNoDKGPrivateInfo insert a record in database to indicate that when DKG was completed, the node failed DKG, there is no DKG key generated.
+func (k *DKGKeys) InsertNoDKGPrivateInfo(epochCounter uint64) error {
+	return operation.RetryOnConflictTx(k.db, transaction.Update, k.storeTx(epochCounter, nil))
+}
+
+// receive the DKG key for the given Epoch, it returns:
+// - (key, true, nil) when DKG was completed, and a DKG private key was saved
+// - (nil, false, nil) when DKG was completed, but no DKG private key was saved
+// - (nil, false, storage.ErrNotFound) no DKG private key is found in database, it's unknown whether a DKG key was generated
+// - (nil, false, error) for other exceptions
+func (k *DKGKeys) RetrieveMyDKGPrivateInfo(epochCounter uint64) (*dkg.DKGParticipantPriv, bool, error) {
 	tx := k.db.NewTransaction(false)
 	defer tx.Discard()
-	return k.retrieveTx(epochCounter)(tx)
+	priv, err := k.retrieveTx(epochCounter)(tx)
+	if err != nil {
+		return nil, false, err
+	}
+	if priv == nil {
+		return nil, false, nil
+	}
+	return priv, true, nil
```

Comment:
error is used for exception. 

If there is no exception, then there are two cases: has or not have the private key. 

I don't like to ask caller to differentiate these by checking `nil`, because if they forgot, and called any methods directly on the nil private key, the complier won't complain and will panic in runtime. Whereas, an extra bool return value will force you to handle it, otherwise won't compile.

Reply:
@tarakby this pattern is similar to go's map lookup 
```golang
m := map[string]float64{"pi": 3.14}
v, found := m["pi"] // v == 3.14  found == true
```
it transforms the `nil` check into a boolean check ... semantically different ... helps people who don't like `nil`s 😉